kratos/processes/process.h

```diff
@@ -253,6 +254,26 @@ class Process : public Flags
 ///@name Type Definitions
 ///@{
 
+#ifndef KRATOS_REGISTER_PROCESS
+#define KRATOS_REGISTER_PROCESS(                                                                             \
+    module_name,                                                                                             \
+    process_name,                                                                                            \
+    process_prototype)                                                                                       \
+    {                                                                                                        \
+        std::string all_path = std::string("Processes.All.") + process_name;                                 \
+        if (!Registry::HasItem(all_path)) {                                                                  \
+            auto& r_process_item = Registry::AddItem<RegistryItem>(all_path);                                \
+            r_process_item.AddItem<RegistryValueItem<Process>>("Prototype", process_prototype);              \
+        } else {                                                                                             \
+            KRATOS_ERROR << "Process '" << process_name << "' is already registered." << std::endl;          \
+        }                                                                                                    \
+        std::string module_path = std::string("Processes.") + module_name + std::string(".") + process_name; \
+        if (!Registry::HasItem(module_path)) {                                                               \
+            auto& r_process_item = Registry::AddItem<RegistryItem>(module_path);                             \
+            r_process_item.AddItem<RegistryValueItem<Process>>("Prototype", process_prototype);              \
+        }                                                                                                    \
+    }
```

Comment:
About this, I'm not getting your point at all about "manual registering"... You mean adding this line to the pointer definition macro or to the already existing one? Because the latter would still require to call it somewhere.

Reply:
You mean something of this sort 
```c++
template<typename TPrototypeType>
void RegisterProcessWithPrototype(
    const std::string RegistryEntryName,
    TPrototypeType& rProcessPrototype)
{
    if (!Registry::HasItem(RegistryEntryName)) {
        auto& r_process_item = Registry::AddItem<RegistryItem>(RegistryEntryName);
        r_process_item.AddItem<RegistryValueItem<TPrototypeType>>("Prototype", rProcessPrototype);
    } else {
        KRATOS_ERROR << "Process '" << RegistryEntryName << "' is already registered." << std::endl;
    }
}

#ifndef KRATOS_REGISTER_PROCESS_WITH_PROTOTYPE
#define KRATOS_REGISTER_PROCESS_WITH_PROTOTYPE(                                                              \
    module_name,                                                                                             \
    process_name,                                                                                            \
    process_prototype)                                                                                       \
    {                                                                                                        \
        std::string all_path = std::string("Processes.All.") + process_name;                                 \
        RegisterProcessWithPrototype(all_path, process_prototype);                                           \
        std::string module_path = std::string("Processes.") + module_name + std::string(".") + process_name; \
        RegisterProcessWithPrototype(module_path, process_prototype);                                        \
    }
#endif
```

Reply:
Yes in this line! Something like:

```c++
template<typename TPrototypeType>
void RegisterProcessWithPrototype(
    const std::string RegistryEntryName,
    TPrototypeType rProcessPrototype)
{
    if (!Registry::HasItem(RegistryEntryName)) {
        auto& r_process_item = Registry::AddItem<RegistryItem>(RegistryEntryName);
        r_process_item.AddItem<RegistryValueItem<TPrototypeType>>("Prototype", rProcessPrototype);
    } else {
        KRATOS_ERROR << "Process '" << RegistryEntryName << "' is already registered." << std::endl;
    }
}

template<typename TPrototypeType>
void RegisterProcessWithPrototype(
    const std::string ModuleName
    const std::string RegistryEntryName,
    TPrototypeType rProcessPrototype)
{
        std::string all_path = std::string("Processes.All.") + process_name;
        RegisterProcessWithPrototype(all_path, process_prototype);
        std::string module_path = std::string("Processes.") + module_name + std::string(".") + process_name;
        RegisterProcessWithPrototype(module_path, process_prototype);    
}

#ifndef KRATOS_PROCESS_DEFAULT_BODY
#define KRATOS_PROCESS_DEFAULT_BODY(                                                               \
    module_name,                                                                                   \
    process)                                                                                       \
    purblic:                                                                                       \
    virtual Process::Pointer Create(Model& rThisModel, Parameters TheParameters) {                 \
        return make_shared<process>(rThisModel, TheParameters);                                    \
    }                                                                                              \
    private:                                                                                       \
    static auto& mrRegistryItem = RegisterProcessWithPrototype(module_name, #process, process());   
    #endif


class AnyProcess{
    private:
    AnyProcess() // Private constructor that we may even add it to the same macro
    KRATOS_PROCESS_DEFAULT_BODY("any_module", AnyProcess)
}
```

Reply:
After giving it a try I see two limitations:
- I'm not sure if it is technically possible as it implies using the default constructor before the class is completed.
- In case we have a template process, as far as I understand, we'll need to register each of the possible specializations. Hence, it is impossible to do it in the header.

Reply:
Finally I decided to create an `RegistryAuxiliary` static class. In here we place the functions above, which are to be called later on in the corresponding `register_processes.cpp` and `register_operations.cpp`.